src/main/kotlin/com/sourcegraph/cody/edit/EditCodeInlayRenderer.kt

```diff
@@ -0,0 +1,61 @@
+package com.sourcegraph.cody.edit
+
+import com.intellij.openapi.editor.EditorCustomElementRenderer
+import com.intellij.openapi.editor.Inlay
+import com.intellij.openapi.editor.markup.TextAttributes
+import com.intellij.ui.JBColor
+import java.awt.Graphics
+import java.awt.Rectangle
+import javax.swing.CellRendererPane
+import javax.swing.JPanel
+
+class EditCodeInlayRenderer(private val component: JPanel, private val requiredHeight: Int) :
+    EditorCustomElementRenderer {
+
+  private val rendererPane = CellRendererPane()
+
+  init {
+    // Call validate() initially to ensure the layout is updated
+    rendererPane.add(component)
+    component.validate()
+  }
+
+  override fun calcWidthInPixels(inlay: Inlay<*>): Int {
+    // TODO: compute the width based on the line width or editor viewport width
+    return maxOf(component.preferredSize.width, MAX_EDITOR_WIDTH)
+  }
+
+  override fun calcHeightInPixels(inlay: Inlay<*>): Int {
+    return requiredHeight
+  }
+
+  override fun paint(
+      inlay: Inlay<*>,
+      g: Graphics,
+      targetRegion: Rectangle,
+      textAttributes: TextAttributes
+  ) {
+    // Various failed attempts to get something to render in the inlay area.
+    rendererPane.bounds = targetRegion
+    component.setBounds(0, 0, targetRegion.width, targetRegion.height)
```

Comment:
Probably won't be the case in your situation, but please double check CPU consumption when changing bounds or recalculating layout in `paint` or `repaint` function.
I introduced [nasty bug which spiked our CPU usage](https://github.com/sourcegraph/jetbrains/pull/271/files) significantly in past because of this.

Reply:
I'm unsure where this is used, but I assume `component` isn't part of the layout, it is just being used to paint into the renderer pane. So laying it out here should have less potential for mayhem.